[SPARK-20331][SQL] Enhanced Hive partition pruning predicate pushdown

mallman opened the pull request:
(Link to Jira: https://issues.apache.org/jira/browse/SPARK-20331)

## What changes were proposed in this pull request?

Spark 2.1 introduced scalable support for Hive tables with huge numbers of partitions. Key to leveraging this support is the ability to prune unnecessary table partitions to answer queries. Spark supports a subset of the class of partition pruning predicates that the Hive metastore supports. If a user writes a query with a partition pruning predicate that is *not* supported by Spark, Spark falls back to loading all partitions and pruning client-side. We want to broaden Spark's current partition pruning predicate pushdown capabilities.

One of the key missing capabilities is support for disjunctions. For example, for a table partitioned by date, writing a query with a predicate like

    date = 20161011 or date = 20161014

will result in Spark fetching all partitions. For a table partitioned by date and hour, querying a range of hours across dates can be quite difficult to accomplish without fetching all partition metadata.

The current partition pruning support supports only comparisons against literals. We can expand that to foldable expressions by evaluating them at planning time.

We can also implement support for the "IN" comparison by expanding it to a sequence of "OR"s.

## How was this patch tested?

The `HiveClientSuite` and `VersionsSuite` were refactored and simplified to make Hive client-based, version-specific testing more modular and conceptually simpler. There are now two Hive test suites: `HiveClientSuite` and `HivePartitionFilteringSuite`. These test suites have a single-argument constructor taking a `version` parameter. As such, these test suites cannot be run by themselves. Instead, they have been bundled into "aggregation" test suites which run each suite for each Hive client version. These aggregation suites are called `HiveClientSuites` and `HivePartitionFilteringSuites`. The `VersionsSuite` and `HiveClientSuite` have been refactored into each of these aggregation suites, respectively.

`HiveClientSuite` and `HivePartitionFilteringSuite` subclass a new abstract class, `HiveVersionSuite`. `HiveVersionSuite` collects functionality related to testing a single Hive version and overrides relevant test suite methods to display version-specific information.

A new trait, `HiveClientVersions`, has been added with a sequence of Hive test versions.

Commit:
[SPARK-20331][SQL] Enhanced partition pruning pushdown

Add support for evaluating foldable, deterministic expressions when
computing Hive metastore filter string for getting filtered partition
list

SparkQA commented:
**[Test build #75783 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75783/testReport)** for PR 17633 at commit [`1a7663d`](https://github.com/apache/spark/commit/1a7663d8c40dc86c59846b52dba9766859d8e095).

SparkQA commented:
**[Test build #75783 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75783/testReport)** for PR 17633 at commit [`1a7663d`](https://github.com/apache/spark/commit/1a7663d8c40dc86c59846b52dba9766859d8e095).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75783/
Test PASSed.

rxin commented:
Does this work for non-Hive tables?

mallman commented:
> Does this work for non-Hive tables?

This is geared towards Hive partitioned tables. If we have another system that prunes table partitions based on a string-ified pruning predicate I'm unaware. Do you have one in mind?

rxin commented:
Then it should work.

tejasapatil commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,34 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
-        s"""${quoteStringLiteral(v.toString)} ${op.symbol} ${a.name}"""
-    }.mkString(" and ")
+    def isFoldable(expr: Expression): Boolean =
+      (expr.dataType.isInstanceOf[IntegralType] || expr.dataType.isInstanceOf[StringType]) &&
```
Can this support all `AtomicType`'s ? From my understanding these are partition columns and can support other types besides int and string.

mallman replied:
`IntegralType` encompasses all "integral" types, including `IntegerType`, `ByteType`, `ShortType`, etc.

I'm trying to be somewhat conservative in what we support here to ensure compatibility. Is there a particular type you'd like to see supported?

mallman commented:
@cloud-fan @ericl Hi guys. Care to review?

ericl commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,34 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+          s"(${convert(expr1)} or ${convert(expr2)})"
+      }
+
+    filters.flatMap(f => Try(convert(f)).toOption).mkString(" and ")
```
Why do we need a `Try` here?

mallman replied:
The `convert` method throws a `MatchError` if an expression is not a compatible Hive partition filter. Otherwise it returns a compatible partition pruning string. While using exception handling in this way can certainly be considered an anti-pattern, it does make for a much simpler implementation and type signature of the `convert` method.

cloud-fan replied:
shall we just make `convert` a `PartialFunction`?

mallman replied:
Seems reasonable. Will do.

mallman replied:
Done.

viirya replied:
So we can get rid of `Try` now?

cloud-fan replied:
for partial function, now we can do `filters.flatMap(f => convert.lift(f))`

viirya replied:
or `filters.collect(convert)`. I've not tried, but should work.

cloud-fan replied:
ah that's better.

mallman replied:
Both `filters.flatMap(f => convert.lift(f))` and `filters.collect(convert)` throw a `MatchError` on input

```
ds=(20170101 + 1) and h=0
```

Now I know that in practice foldable expressions such as `20170101 + 1` are converted to literals, but the point I'm making here is that the alternatives proposed here will throw a `MatchError` if the filter expression is parsed into a tree with a leaf that does not match a pattern defined by the `convert` partial function.

Hence the use of `filters.flatMap(f => Try(convert(f)).toOption).mkString(" and ")`.

gatorsmile commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,34 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+      expr.foldable &&
+      expr.deterministic
+
+    def convertFoldable(expr: Expression): String = expr.dataType match {
```
The foldable expressions should be converted in Optimizer, right?

cloud-fan replied:
yea, to make the code simpler, we can just match literals.

gatorsmile commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,34 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+    def convert(filter: Expression): String =
+      filter match {
+        case In(a: Attribute, exprs) if exprs.forall(isFoldable) =>
+          val or = exprs.map(expr => s"${a.name} = ${convertFoldable(expr)}").reduce(_ + " or " + _)
```
No need to quote the column names?

mallman replied:
Good point. You mean with backticks, like

    s"`${a.name}` = ${convertFoldable(expr)}"

?

cloud-fan replied:
see `AttributeReference.sql`, we should use `quoteIdentifier`

mallman replied:
In testing this in a modified form of `HiveClientSuite`, Hive is complaining that it can't parse the predicate. Specifically, the error message in the exception it's throwing is

```
Error parsing partition filter : line 1:3 no viable alternative at character '`'
```

The filter string it's trying to parse is `` `ds` = 20170101``.

gatorsmile commented:
Could you check whether there exists any limit on predicate we can pass to Hive? Are they the same for all our supported Hive megastore versions?

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala:
```diff
@@ -43,19 +47,159 @@ class HiveClientSuite extends SparkFunSuite {
     assert(filteredPartitions.size == testPartitionCount)
   }
+
+  test("getPartitionsByFilter: ds=20170101") {
```
will these tests be executed on all supported hive versions?

mallman replied:
> will these tests be executed on all supported hive versions?

No. That's something I'll look into.

mallman replied:
They will now. See `HiveClientSuites.scala` in this PR.

mallman commented:
Hi Guys. Sorry for the lack of updates on this. I've been held up with other responsibilities the past week. I'm planning to push a new commit today or tomorrow.

Commit:
Remove support for "foldables" from the Hive metastore partition pruning
predicate constructor

SparkQA commented:
**[Test build #76706 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76706/testReport)** for PR 17633 at commit [`a4cdfb0`](https://github.com/apache/spark/commit/a4cdfb05d3a0ba154b2e3fe17b48133cd0563e0c).

mallman commented:
I've pushed a new commit removing the logical for handling "foldables", since these are evaluated earlier in planning.

I've also removed the modifications I made to `FiltersSuite.scala` and modified the `convertFilters` method to conform to its original specification.

mallman commented:
> Could you check whether there exists any limit on predicate we can pass to Hive?

There are, and I found something in the way of documentation or a grammar a while back that specifies the accepted predicates. I'll dig that up.

> Are they the same for all our supported Hive megastore versions?

I don't know yet. I have cross-version compatibility checking on my to-do list.

SparkQA commented:
**[Test build #76706 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76706/testReport)** for PR 17633 at commit [`a4cdfb0`](https://github.com/apache/spark/commit/a4cdfb05d3a0ba154b2e3fe17b48133cd0563e0c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76706/
Test PASSed.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,39 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
-    }.mkString(" and ")
+    def isExtractable(expr: Expression): Boolean =
+      expr match {
+        case Literal(_, _: IntegralType) | Literal(_, _: StringType) => true
```
what about float/double/decimal?

mallman replied:
I'm going to look into support for that. FWIW, it's not supported in the current codebase, so omitting it wouldn't be a regression.

mallman replied:
I suggest we omit support for this at this time to reduce the necessary testing and reviewing footprint. We can add support for other data types as part of a new Jira issue and PR. Okay?

mallman commented:
Hey guys. Just a quick update. I made good progress on implementing multi-version testing today, however it's not quite ready. I'm going to be on leave from tomorrow through the rest of next week, so I'm kind of doubtful I'll push anything new until May 22nd.

gatorsmile commented:
You can add the test cases to `VersionsSuite` for verifying the behaviors of each Hive meta-store client version.

SparkQA commented:
**[Test build #77199 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77199/testReport)** for PR 17633 at commit [`4f802a5`](https://github.com/apache/spark/commit/4f802a5c955bdcf6a9387de0aa37f8bc70a318cd).

mallman commented:
> You can add the test cases to VersionsSuite for verifying the behaviors of each Hive meta-store client version.

@gatorsmile, I've refactored Hive version-specific testing to make it easier to add new Hive test suites which test functionality for a collection of Hive versions. I did not feel adding these partition filtering tests to `VersionsSuite.scala` was a good fit.

mallman commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala:
```diff
@@ -17,45 +17,656 @@
+    assert("success" === client.getConf("test", null))
+  }
 
+  // Should this be in a beforeAll() method instead?
```
Should this be in a `beforeAll()` method instead? This actually initializes `client` for use by other tests. Putting that in a `beforeAll()` method feels more appropriate.

AmplabJenkins commented:
Merged build finished. Test FAILed.

mallman commented:
I think this build was aborted because of the emergency jenkins restart, as reported on the spark dev mailing list. Retest, please?

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #77242 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77242/testReport)** for PR 17633 at commit [`4f802a5`](https://github.com/apache/spark/commit/4f802a5c955bdcf6a9387de0aa37f8bc70a318cd).

SparkQA commented:
**[Test build #77242 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77242/testReport)** for PR 17633 at commit [`4f802a5`](https://github.com/apache/spark/commit/4f802a5c955bdcf6a9387de0aa37f8bc70a318cd).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `class HiveClientSuite(version: String) extends HiveVersionSuite(version) with HiveClientVersions `
  * `class HiveClientSuites extends Suite with HiveClientVersions `
  * `class HivePartitionFilteringSuite(version: String)`
  * `class HivePartitionFilteringSuites extends Suite with HiveClientVersions `

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77242/
Test PASSed.

mallman commented:
Rebased to resolve merge conflicts.

SparkQA commented:
**[Test build #77262 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77262/testReport)** for PR 17633 at commit [`0c4040c`](https://github.com/apache/spark/commit/0c4040c1676f5db702b2427651851027703e310d).

SparkQA commented:
**[Test build #77262 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77262/testReport)** for PR 17633 at commit [`0c4040c`](https://github.com/apache/spark/commit/0c4040c1676f5db702b2427651851027703e310d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class HiveClientSuite(version: String) extends HiveVersionSuite(version) with HiveClientVersions `
  * `class HiveClientSuites extends Suite with HiveClientVersions `
  * `class HivePartitionFilteringSuite(version: String)`
  * `class HivePartitionFilteringSuites extends Suite with HiveClientVersions `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77262/
Test PASSed.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,39 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+
+    def convert(filter: Expression): String =
+      filter match {
+        case In(a: Attribute, exprs) if exprs.forall(isExtractable) =>
```
shall we check `varcharKeys` here?

mallman replied:
Good catch. Will do.

cloud-fan commented:
it's really hard to review the tests... Can we just add some simple tests and refactor the test suites in a follow-up PR?

SparkQA commented:
**[Test build #77306 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77306/testReport)** for PR 17633 at commit [`a1107ae`](https://github.com/apache/spark/commit/a1107ae0722d10c948523fd98c20425e90d823f3).

mallman commented:
> it's really hard to review the tests... Can we just add some simple tests and refactor the test suites in a follow-up PR?

It's a big, complicated diff, and I sympathize. Basically, I started adding support for testing this new functionality across Hive versions and saw how much code duplication there was with the existing `VersionsSuite.scala`. And I didn't want to just jam all the new tests into `VersionsSuite.scala`. The new tests really are rather specialized and belong in their own test suite. So I refactored as I described in my (updated) PR description to create something that reduced code duplication and laid the foundation for additional, specialized, cross-Hive-version tests.

I think as a compromise we might leave the existing `VersionsSuite.scala` code alone, acknowledge the code duplication with the new test code and follow up with a complete refactoring. I think that would simplify the review of the new tests without sacrificing test coverage.

SparkQA commented:
**[Test build #77306 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77306/testReport)** for PR 17633 at commit [`a1107ae`](https://github.com/apache/spark/commit/a1107ae0722d10c948523fd98c20425e90d823f3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77306/
Test PASSed.

cloud-fan commented:
@mallman yea that sounds good to me

Commit:
Test Hive table partition filtering by Hive client version

Commit:
Convert the metastore filter expression convert function to a partial
function and add a guard when testing In expressions against varchar
attributes

mallman commented:
> @mallman yea that sounds good to me

Okay. The testing diff is significantly simpler now.

SparkQA commented:
**[Test build #77545 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77545/testReport)** for PR 17633 at commit [`358e69c`](https://github.com/apache/spark/commit/358e69cee2f1963dc3d6abf3dcac7926c1155ec5).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77545/
Test FAILed.

SparkQA commented:
**[Test build #77939 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77939/testReport)** for PR 17633 at commit [`7965ef3`](https://github.com/apache/spark/commit/7965ef35ce45dbaabb7ba525a1b41625365b9da6).

SparkQA commented:
**[Test build #77939 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77939/testReport)** for PR 17633 at commit [`7965ef3`](https://github.com/apache/spark/commit/7965ef35ce45dbaabb7ba525a1b41625365b9da6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77939/
Test PASSed.

mallman commented:
@cloud-fan Somehow I didn't notice that my earlier commit was failing scalastyle. Anyway, I've pushed a new version that passes all tests. Can you continue your review, please?

mallman commented:
@cloud-fan ping

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #79117 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79117/testReport)** for PR 17633 at commit [`7965ef3`](https://github.com/apache/spark/commit/7965ef35ce45dbaabb7ba525a1b41625365b9da6).

gatorsmile commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,40 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+
+    lazy val convert: PartialFunction[Expression, String] =
+      {
+        case In(a: Attribute, exprs)
```
Maybe we should not push down IN, like what we did for parquet?

mallman replied:
Why not?

gatorsmile commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,40 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+      }
+
+    lazy val convert: PartialFunction[Expression, String] =
+      {
```
Please fix the indent.

gatorsmile commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,40 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+            if !varcharKeys.contains(a.name) && exprs.forall(isExtractable) =>
+          val or =
+            exprs
+              .map(expr => s"${a.name} = ${extractValue(expr)}")
```
Do we need to quote the name? How about adding `(` `)` for each condition?

gatorsmile commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,40 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+          "(" + or + ")"
+        case op @ BinaryComparison(a: Attribute, expr2)
+            if !varcharKeys.contains(a.name) && isExtractable(expr2) =>
+          s"${a.name} ${op.symbol} ${extractValue(expr2)}"
```
Do we need to quote the name? How about adding `(` `)` for each condition?

mallman replied:
Quoting the name causes the Hive metastore partition filter parser to throw an exception.

I don't think we need to add any more parens. Can you give a counterexample?

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,40 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+              .map(expr => s"${a.name} = ${extractValue(expr)}")
+              .reduce(_ + " or " + _)
+          "(" + or + ")"
+        case op @ BinaryComparison(a: Attribute, expr2)
```
how about we define a
```
object ExtractableLiteral {
  def unapply(e: Expression): Option[String] = e match {
    case Literal(v, _: IntegralType) => Some(v.toString)
    case Literal(v, _: StringType) => Some(quoteStringLiteral(v.toString))
    case _ => None
  }
}
```
and then we can just write
```
case op @ BinaryComparison(a: Attribute, ExtractableLiteral(value)) if varcharKeys... =>
  s"${a.name} ${op.symbol} $value"
```

viirya replied:
+1 So we can avoid duplicate cases when adding new types.

rxin replied:
while it saves a few lines, it can be really confusing though to define an ExtractableLiteral that implicitly turns an int into a string ...

mallman replied:
I like the idea of an extractor object. Maybe the name is confusing, but I'm at a loss for proposing a better name.

cloud-fan replied:
how about `ExtractLiteralToString`?

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,40 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+        case Literal(v, _: StringType) => quoteStringLiteral(v.toString)
+      }
+
+    lazy val convert: PartialFunction[Expression, String] =
```
I think it's ok to not use lazy val here, creating a partition function is cheap. BTW, we should use this style:
```
val convert: PartialFunction[Expression, String] = {
  case ...
}
```

mallman replied:
I don't think you can define a recursive partial function without making it lazy.

viirya commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,40 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+            if !varcharKeys.contains(a.name) && isExtractable(expr1) =>
+          s"${extractValue(expr1)} ${op.symbol} ${a.name}"
+        case op @ And(expr1, expr2) =>
+          s"(${convert(expr1)} and ${convert(expr2)})"
```
If expr1 or expr2 is not defined at `convert`, does this work? I think either `filters.flatMap(f => convert.lift(f))` or `filters.collect(convert)` will encounter scala.MatchError, right?

cloud-fan replied:
we can follow `InMemoryTableScanExec.buildFilter`

mallman replied:
@viirya This is why we wrap the call to `convert` in a `Try` at the end of the method.

viirya commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,40 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+        case op @ And(expr1, expr2) =>
+          s"(${convert(expr1)} and ${convert(expr2)})"
+        case op @ Or(expr1, expr2) =>
+          s"(${convert(expr1)} or ${convert(expr2)})"
```
ditto.

mallman replied:
See above.

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuites.scala:
```diff
@@ -0,0 +1,29 @@
+
+class HiveClientSuites extends Suite with HiveClientVersions {
+  override def nestedSuites: IndexedSeq[Suite] = {
+    // Hive 0.12 does not provide the partition filtering API we call
```
then we will lost test coverage because we don't test hive 0.12 for basic operations...  How about we create a new `HivePartitionFilteringSuite` and move the partition related tests there?

mallman replied:
No test coverage has been lost. `HiveClientSuite` is not a good name for that class. It actually just tests partition filtering. I added a `TODO` suggesting we rename it to `HivePartitionFilteringSuite`, but that makes the diff difficult to grok. All other Hive client testing is still in `VersionsSuite`, and that suite still tests all supported Hive versions. I wanted to change the names of these tests to make them clearer, but it resulted in a jumbled diff. You requested that I keep the names as-is to make the diff clearer.

I added "TODO" comments suggesting better names for these classes, for renaming later.

SparkQA commented:
**[Test build #79117 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79117/testReport)** for PR 17633 at commit [`7965ef3`](https://github.com/apache/spark/commit/7965ef35ce45dbaabb7ba525a1b41625365b9da6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79117/
Test PASSed.

Commit:
Add ExtractableLiteral and ExtractableLiterals, fix some indentation and
add another test case

SparkQA commented:
**[Test build #79300 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79300/testReport)** for PR 17633 at commit [`d88b8ab`](https://github.com/apache/spark/commit/d88b8ab091a173cb88ad4db99d2a4b0cc2b0921e).

SparkQA commented:
**[Test build #79300 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79300/testReport)** for PR 17633 at commit [`d88b8ab`](https://github.com/apache/spark/commit/d88b8ab091a173cb88ad4db99d2a4b0cc2b0921e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79300/
Test PASSed.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,43 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+    }
+
+    lazy val convert: PartialFunction[Expression, String] = {
+      case In(a: Attribute, ExtractableLiterals(values)) if !varcharKeys.contains(a.name) =>
```
cc @gatorsmile , any concerns to not do it?

viirya replied:
Shall we also consider `InSet`?

mallman replied:
I'll look into it.

mallman replied:
Done, though please double check the implementation.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +590,43 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+        s"(${convert(expr1)} or ${convert(expr2)})"
+    }
+
+    filters.flatMap(f => Try(convert(f)).toOption).mkString(" and ")
```
I do think we should follow `InMemoryTableScanExec.buildFilters`. For example, if the left side of `And` is not supported but the right side is, and we can still push down the right side. But here, we simply catch the exception and push nothing.

mallman replied:
I see your point. I'll look into an alternative, such as the approach taken by `InMemoryTableScanExec.buildFilters`.

mallman replied:
Done.

Commit:
Support nested AND expressions and InSet

SparkQA commented:
**[Test build #79493 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79493/testReport)** for PR 17633 at commit [`a087a0f`](https://github.com/apache/spark/commit/a087a0f44372a226c64514d8a7163801b8eb54ef).

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +591,67 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+
+    object ExtractableLiterals {
+      def unapply(exprs: Seq[Expression]): Option[Seq[String]] = {
+        exprs.map(ExtractableLiteral.unapply).foldLeft(Option(Seq.empty[String])) {
```
I'd like it to be more java style:
```
val extracted = exprs.map(ExtractableLiteral.unapply)
if (extracted. exists(_.isEmpty)) {
  None
} else {
  extracted.map(_.get)
}
```

mallman replied:
Is there something wrong with the way it is now?

cloud-fan replied:
`foldLeft` may be not friendly to some Spark developers, but it's not a big deal.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +591,67 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+      }
+
+      def unapply(values: Set[Any]): Option[Seq[String]] = {
+        values.toSeq.foldLeft(Option(Seq.empty[String])) {
```
ditto

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +591,67 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+        (convert.lift(expr1) ++ convert.lift(expr2)).mkString("(", " and ", ")")
+      case op @ Or(expr1, expr2)
+          if convert.isDefinedAt(expr1) && convert.isDefinedAt(expr2) =>
+        (convert.lift(expr1) ++ convert.lift(expr2)).mkString("(", " or ", ")")
```
nit: s"(convert(expr1) or convert(expr))"

mallman replied:
Ok.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -589,18 +591,67 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
           if !varcharKeys.contains(a.name) =>
-        s"""${a.name} ${op.symbol} ${quoteStringLiteral(v.toString)}"""
-      case op @ BinaryComparison(Literal(v, _: StringType), a: Attribute)
+        s"${a.name} ${op.symbol} $value"
```
shall we add `()` for binary comparisons?

mallman replied:
Is there a problem with leaving them out?

cloud-fan replied:
nvm, realized that `and`, `or` have higher precedence over binary operators, so it should be fine

SparkQA commented:
**[Test build #79493 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79493/testReport)** for PR 17633 at commit [`a087a0f`](https://github.com/apache/spark/commit/a087a0f44372a226c64514d8a7163801b8eb54ef).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79493/
Test PASSed.

Commit:
When dealing with an Or expression where both children are convertible
to a filter string, we can safely call convert(expr) on each directly

SparkQA commented:
**[Test build #79501 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79501/testReport)** for PR 17633 at commit [`af3065a`](https://github.com/apache/spark/commit/af3065a0da64251350eca241b4eb3c3c9447f114).

cloud-fan commented:
LGTM, pending jenkins

SparkQA commented:
**[Test build #79501 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79501/testReport)** for PR 17633 at commit [`af3065a`](https://github.com/apache/spark/commit/af3065a0da64251350eca241b4eb3c3c9447f114).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79501/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

mallman commented:
@cloud-fan Can you back port this PR to 2.1 and 2.2, please? I think the patch should apply cleanly.

rxin commented:
@mallman we don't backport such risky changes to maintenance branches. Those branches typically go through much less testing.

yujhe commented:
hi @mallman 
I can run this test via this command:
`build/sbt "test-only org.apache.spark.sql.hive.client.HiveClientSuites`

But when I try to run this test on Intellij IDEA, I got the following error message, do you have any idea to run this test on Intellij?
```
Exception encountered when invoking run on a nested suite - [download failed: com.google.inject#guice;3.0!guice.jar, download failed: javax.inject#javax.inject;1!javax.inject.jar, download failed: aopalliance#aopalliance;1.0!aopalliance.jar, download failed: com.sun.jersey#jersey-server;1.9!jersey-server.jar(bundle), download failed: com.sun.jersey#jersey-json;1.9!jersey-json.jar(bundle), download failed: org.codehaus.jettison#jettison;1.1!jettison.jar(bundle), download failed: com.sun.xml.bind#jaxb-impl;2.2.3-1!jaxb-impl.jar, download failed: com.sun.jersey.contribs#jersey-guice;1.9!jersey-guice.jar]
java.lang.RuntimeException: [download failed: com.google.inject#guice;3.0!guice.jar, download failed: javax.inject#javax.inject;1!javax.inject.jar, download failed: aopalliance#aopalliance;1.0!aopalliance.jar, download failed: com.sun.jersey#jersey-server;1.9!jersey-server.jar(bundle), download failed: com.sun.jersey#jersey-json;1.9!jersey-json.jar(bundle), download failed: org.codehaus.jettison#jettison;1.1!jettison.jar(bundle), download failed: com.sun.xml.bind#jaxb-impl;2.2.3-1!jaxb-impl.jar, download failed: com.sun.jersey.contribs#jersey-guice;1.9!jersey-guice.jar]
	at org.apache.spark.deploy.SparkSubmitUtils$.resolveMavenCoordinates(SparkSubmit.scala:1303)
	at org.apache.spark.sql.hive.client.IsolatedClientLoader$$anonfun$2.apply(IsolatedClientLoader.scala:111)
	at org.apache.spark.sql.hive.client.IsolatedClientLoader$$anonfun$2.apply(IsolatedClientLoader.scala:111)
	at org.apache.spark.sql.catalyst.util.package$.quietly(package.scala:42)
	at org.apache.spark.sql.hive.client.IsolatedClientLoader$.downloadVersion(IsolatedClientLoader.scala:110)
	at org.apache.spark.sql.hive.client.IsolatedClientLoader$.liftedTree1$1(IsolatedClientLoader.scala:61)
	at org.apache.spark.sql.hive.client.IsolatedClientLoader$.forVersion(IsolatedClientLoader.scala:60)
	at org.apache.spark.sql.hive.client.HiveClientBuilder$.buildClient(HiveClientBuilder.scala:50)
	at org.apache.spark.sql.hive.client.HiveVersionSuite.buildClient(HiveVersionSuite.scala:39)
	at org.apache.spark.sql.hive.client.HiveClientSuite.org$apache$spark$sql$hive$client$HiveClientSuite$$init(HiveClientSuite.scala:48)
	at org.apache.spark.sql.hive.client.HiveClientSuite.beforeAll(HiveClientSuite.scala:70)
	at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:212)
	at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:210)
	at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:31)
	at org.scalatest.Suite$class.callExecuteOnSuite$1(Suite.scala:1210)
	at org.scalatest.Suite$$anonfun$runNestedSuites$1.apply(Suite.scala:1257)
	at org.scalatest.Suite$$anonfun$runNestedSuites$1.apply(Suite.scala:1255)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at org.scalatest.Suite$class.runNestedSuites(Suite.scala:1255)
	at org.apache.spark.sql.hive.client.HiveClientSuites.runNestedSuites(HiveClientSuites.scala:24)
	at org.scalatest.Suite$class.run(Suite.scala:1144)
	at org.apache.spark.sql.hive.client.HiveClientSuites.run(HiveClientSuites.scala:24)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:45)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$1.apply(Runner.scala:1340)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$1.apply(Runner.scala:1334)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1334)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1011)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1010)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1500)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1010)
	at org.scalatest.tools.Runner$.run(Runner.scala:850)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:131)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:28)
```
